### ISSUE TITLE
CancellationToken support

### ISSUE THREAD

**Issue by UnoSD** [Azure/azure-service-bus-dotnet#258]
## Actual Behavior
ReceiveAsync and ReceiveBatchAsync wait until a message is present on the queue. To shut down, the Abort() method must be invoked or a timeout added (which will lead to pointless loops); the Abort() will abort all the receive calls, the only way around is breaking the guidelines of the lifetime and create multiple clients in the application for the same queue (not tested and might fail if the internal cache of queues returns the same).

## Expected Behavior
An overload of the receive methods supporting a cancellation token to exit gracefully a call.

Thank you.

**Comment by SeanFeldman** [comment]
[.NET client](https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.messaging.queueclient.receiveasync) didn't have that overload as well.

[`RegisterMessageHandler`](https://github.com/Azure/azure-service-bus-dotnet/blob/7cd009cc2daf90105025d5cf6d9e80fa203a5c63/src/Microsoft.Azure.ServiceBus/QueueClient.cs#L394) accepts a token. Receive async or receive batch async do expose a an overload with `TimeSpan` and are intended for simpler scenarios.

**Comment by UnoSD** [comment]
`RegisterMessageHandler` unfortunately does not support batch. 

I am currently using the timeout, but: a short timeout causes the service to loop unnecessarily several time, a long timeout then delays heavily the shutdown as it needs to wait on the expiration (we may solve it using `.Abort()` if it breaks the wait).

Would be nice to see this feature, I would create an issue on the .NET Client repository if it is on GitHub.

**Comment by SeanFeldman** [comment]
.NET client most likely won't get new feature as it's planned to be in
supported for critical  bugs only. If you want this feature for the new
client (netstd), this issue is it 🙂

**Comment by UnoSD** [comment]
Thank you :)

**Comment by SeanFeldman** [comment]
I'm not to be thanked, unless @Azure/azure-service-bus-write decide to mark it as a feature request and it's eventually implemented 🙂

**Comment by springcomp** [comment]
I'm migrating code from the .Net full client to the .Net standard client, using the [`MessageReceiver.ReceiveAsync`](https://github.com/Azure/azure-service-bus-dotnet/blob/d13a65a4d93faa09a7d94fae89359b8ad389a489/src/Microsoft.Azure.ServiceBus/Core/IMessageReceiver.cs) method with a long operationTimeout value in a loop.
Because the `ReceiveAsync()` method does not accept a [`CancellationToken`](https://msdn.microsoft.com/en-us/library/system.threading.cancellationtoken(v=vs.110).aspx),
I chose to use the following code, to instruct the receiver to stop listening to the service bus entity.

    while (!cancellationToken.IsCancellationRequested)
    {
        using (cancellationToken.Register(async () =>
        {
            // this should break the .ReceiveAsync() call below
            await receiver_.CloseAsync();
        }))

        {
            try
            {
                var message = await receiver_.ReceiveAsync(operationTimeout);
                if (cancellationToken.IsCancellationRequested)
                    break;

                if (message == null)
                    continue;

                // ... processing message
            }
            catch (MessagingEntityNotFoundException e)
            {
            }
            catch (MessagingEntityDisabledException e)
            {
            }
            catch (ObjectDisposedException)
            {
                // this is expected
            }
        }
    }

I found that the `CloseAsync()` code does not break the `ReceiveAsync()` call.

I investigated the source code of `MessageReceiver` and found that, in the protected `OnReceiveAsync()` method, the main loop that receives messages using an AMQP Receive Link keeps on running even though the Receive Link Manager has been closed. I wonder if that is a bug in the dependent AMQP library.

However, I found that I could make my code work by modifying the `OnReceiveAsync()` call by introducing a check for `IsClosedOrClosing` at each iteration of the loop, like so:

    while (timeoutHelper.RemainingTime() > TimeSpan.Zero)
    {
        if (IsClosedOrClosing)
            throw new ObjectDisposedException(Resources.ReceiverIsClosedOrClosing, (Exception)null);

        bool hasMessages = await Task.Factory.FromAsync(
        (c, s) => receiveLink.BeginReceiveRemoteMessages(maxMessageCount, DefaultBatchFlushInterval, timeoutHelper.RemainingTime(), c, s),
        a => receiveLink.EndReceiveMessages(a, out amqpMessages),
        this).ConfigureAwait(false);

        ...
    }

I would like to know if the maintainers thinks this should be the intended behavior.
If not, would they be happy to let me propose a pull request with a reasonable chance of being accepted?

**Comment by UnoSD** [comment]
@springcomp `CloseAsync` and `Abort` (I think Abort does what you are after, not 100% sure though) are usable only if you have only one caller awaiting `ReceiveAsync`. QueueClient is thread safe so can be used concurrently by many consumers and you may want to cancel only one call.

If you have only one, give `Abort` a try, but for my case is not a solution unfortunately.

**Comment by springcomp** [comment]
@UnoSD I cannot seem to find a function named `Abort`. The only reference I can find is on a class named `AmqpLinkCreator` which seems low level to me. (I'm looking at the code as from [this]([url](https://github.com/Azure/azure-service-bus-dotnet/search?utf8=%E2%9C%93&q=51f8d77f0457&type=Commits)) commit).

I'm not sure what you are referring to. If I create multiple instances of the `MessageReceiver` class as instructed by [this sample](https://github.com/Azure/azure-service-bus/tree/master/samples/DotNet/Microsoft.Azure.ServiceBus/SendReceiveUsingMessageSenderReceiver), I'm pretty confident each one can be `CloseAsync`-ed independently, no ?

**Comment by UnoSD** [comment]
@springcomp Probably is not available on the .NET Standard version.

https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.messaging.cliententity.abort?view=azureservicebus-4.1.1#Microsoft_ServiceBus_Messaging_ClientEntity_Abort

I can't find where it states you have to create multiple instances in the link, but the .NET Full client recommended to create only one instance and reuse it (for several reasons, one of which is a bug I spotted that causes a memory leak as ClientEntities are cached on creation and not removed on disposal); it may even have some other optimization I am not aware of, but I am talking about the full framework client, I have not spent time testing/exploring the .NET Standard one.

**Comment by nemakam** [comment]
@springcomp
Your approach 
```c#
while (timeoutHelper.RemainingTime() > TimeSpan.Zero)
{
    this.ThrowIfClosed();

    bool hasMessages = await Task.Factory.FromAsync(
    (c, s) => receiveLink.BeginReceiveRemoteMessages(maxMessageCount, DefaultBatchFlushInterval, timeoutHelper.RemainingTime(), c, s),
    a => receiveLink.EndReceiveMessages(a, out amqpMessages),
    this).ConfigureAwait(false);

    ...
}
```
looks fine. You can send a PR for that if that's blocking you. But right fix would be is to have  `receiveLink.BeginReceiveRemoteMessages()` throw an exception when the link is closed in between. That's added as a new item in the backlog for amqp library team. Meanwhile, your approach looks good. (y)

**Comment by nemakam** [comment]
@UnoSD , the expectation is that `await queueClient.CloseAsync()` should close ASAP and not wait for the timeout to finish. Looks like this is a bug in our code. 
For your scenario, the right approach is to use a single `queueClient` with probably the default operation timeout, and call `await queueClient.CloseAsync()` when you need to shut down.

**Comment by nemakam** [comment]
Closed via #330

**Comment by UnoSD** [comment]
@nemakam can this be re-opened? Even if fixed `CloseAsync` will break all the `Receive` calls and users could have multiple calls waiting for a message and not break all at them at the same time.

**Comment by nemakam** [comment]
> users could have multiple calls waiting for a message and not break all at them at the same time.

@UnoSD I didnt get what you mean by this. Could you explain it further? Probably answering "Actual Behaviour" and "Expected Behavior" will make it clear.

**Comment by UnoSD** [comment]
**Actual behavior**

```csharp
// Let's pretend those are two services processing messages from
// the same queue in parallel.
await Task.WhenAll(client.ReceiveAsync(), client.ReceiveAsync());

// This breaks both the calls above.
await client.CloseAsync();
```

**Expected behavior**

```csharp
var first = new CancellationTokenSource();
var second = new CancellationTokenSource();

await Task.WhenAll(client.ReceiveAsync(first.Token), client.ReceiveAsync(second.Token));

// Shutting down service one only
first.Cancel();
```

**Comment by nemakam** [comment]
@UnoSD 
I understand your requirement now. The way to go ahead with this is to create a new `QueueClient` for each of the service. 

> I can't find where it states you have to create multiple instances in the link, but the .NET Full client recommended to create only one instance and reuse it (for several reasons, one of which is a bug I spotted that causes a memory leak as ClientEntities are cached on creation and not removed on disposal); it may even have some other optimization I am not aware of, but I am talking about the full framework client, I have not spent time testing/exploring the .NET Standard one.

So the main purpose of reusing client is to avoid creating new connections for every operation (we see many customers take this approach which is not a good idea). If you want independence on when you want to close a client, then creating a new client for each service is the right way to go. As long as you call `QueueClient.Close`, there should not be any memory leaks.

**Comment by UnoSD** [comment]
That would work if they're not cached internally (I vaguely recall a collection of ClientEntities stored in the full framework client). Although we do use a big number of clients (cached at the moment) and that would possibly affect memory/performances.

**Comment by nemakam** [comment]
@UnoSD 
How big is `big number of clients`?

Either way, once a client has been closed, nothing should be leaked in memory. If that's not happening, you can raise a bug and we will look into it. But for your scenario, you should definitely consider having multiple clients.

**Comment by UnoSD** [comment]
I will check as I'm not sure that will actually matter, but we have several services running on a single process each using a queue and more than one using the same often.

Plus our IIS process that needs a queue on each request, but being send only, in that case we can probably get away with caching.

We will try and resolve it with multiple clients and `CloseAsync`, but I have a feeling that people in other scenarios may also benefit from this more than us.

Implementation-wise, I think it makes it also more messy to register the token we have to invoke `CloseAsync` instead of passing it into the call we want to cancel. But if there is no chance it will be implemented we will bite the bullet.

**Comment by nemakam** [comment]
@UnoSD 
Firstly, we will not be adding new functionality to the client you are using, unless its a bug. All new development is happening in `Microsoft.Azure.ServiceBus` client (which is in this repo).
And secondly, if we get other requests for the same feature with a valid scenario, we will take a look at it then. 
For your scenario, I'd suggest going ahead with `CloseAsync()`

**Comment by UnoSD** [comment]
@nemakam of course, I suggested it not strictly for us right now. It's a nice to have when we migrate to the new client and I think many would benefit from it.

Will give updates in case we have troubles using `CloseAsync` and several clients.
